Backport #64767 to 23.8: Fix crash with DISTINCT and window functions

robot-ch-test-poll2 opened the pull request:
Original pull-request https://github.com/ClickHouse/ClickHouse/pull/64767
Cherry-pick pull-request #64996

This pull-request is a last step of an automated backporting.
Treat it as a standard pull-request: look at the checks and resolve conflicts.
Merge it only if you intend to backport changes to the target branch, otherwise just close it.

robot-clickhouse committed:
Backport #64767 to 23.8: Fix crash with DISTINCT and window functions

robot-ch-test-poll commented:
<!-- automatic status comment for PR #64997 from ClickHouse/ClickHouse:backport/23.8/64767 -->
*This is an automated comment for commit 6dfabd796703f0bb479a952560ed610b6c9af2fc with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/64997/6dfabd796703f0bb479a952560ed610b6c9af2fc/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Push to Dockerhub</td><td>The check for building and pushing the CI related docker images to docker hub</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

devcrafter committed:
Update 03165_distinct_with_window_func_crash.sql

Remove query_plan_enable_multithreading_after_window_functions, it doesn't exist in 23.8

devcrafter committed:
Merge remote-tracking branch 'origin/23.8' into backport/23.8/64767
